source/Engine.cpp

```diff
@@ -1420,13 +1421,21 @@ void Engine::CalculateStep()
 	// Check if the flagship just entered a new system.
 	if(flagship && playerSystem != flagship->GetSystem())
 	{
+		bool wormholeEntry = false;
 		// Wormhole travel: mark the wormhole "planet" as visited.
 		if(!wasHyperspacing)
 			for(const auto &it : playerSystem->Objects())
 				if(it.HasValidPlanet() && it.GetPlanet()->IsWormhole() &&
 						&it.GetPlanet()->GetWormhole()->WormholeDestination(*playerSystem) == flagship->GetSystem())
+				{
+					wormholeEntry = true;
 					player.Visit(*it.GetPlanet());
+				}
 
+		player.SetSystemEntry(wormholeEntry ? PlayerInfo::SystemEntry::WORMHOLE :
+			flagship->IsUsingJumpDrive() ? PlayerInfo::SystemEntry::JUMP :
+			PlayerInfo::SystemEntry::HYPERDRIVE);
+		player.Conditions().Set("previous system: " + playerSystem->Name(), 1);
```

Comment:
These shouldn't be set as conditions. There should be two new variables in PlayerInfo, `System *previousSystem` and `SystemEntry enteredSystemBy `. These should be saved out in PlayerInfo::Save.
```c++
	if(system)
		out.Write("system", system->Name());
	if(planet)
		out.Write("planet", planet->TrueName());
+	if(previousSystem)
+		out.Write("previous system", previousSystem->Name());
+	if(enteredSystemBy != SystemEntry::NONE)
+	{
+		// Determine what to write out for the current enteredVia value.
+	}
```
These same variables should also be read in during PlayerInfo::Load
```
else if(child.Token(0) == "previous system" && child.Size() >= 2)
	previousSystem = GameData::Systems().Get(child.Token(1));
else if(child.Token(0) == "entered system by" && child.Size() >= 2)
{
	// Determine what to set the enum to based on the string value.
}
```
PlayerInfo::RegisterDerivedConditions should then have a prefixed provider for each of these.

We don't want to use normal conditions anymore for this type of stuff.

Reply:
Right, wasnt sure how to do it but won't take much work I already have the proper function in playerinfo